src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2816 @@
+use mz_storage_types::connections::ConnectionContext;
+use mz_storage_types::controller::{
+    CollectionMetadata, PersistTxnTablesImpl, StorageError, TxnsCodecRow,
+};
+use mz_storage_types::dyncfgs::STORAGE_DOWNGRADE_SINCE_DURING_FINALIZATION;
+use mz_storage_types::parameters::StorageParameters;
+use mz_storage_types::read_holds::{ReadHold, ReadHoldError};
+use mz_storage_types::read_policy::ReadPolicy;
+use mz_storage_types::sources::{
+    GenericSourceConnection, IngestionDescription, SourceConnection, SourceData, SourceDesc,
+    SourceExport,
+};
+use timely::progress::frontier::MutableAntichain;
+use timely::progress::{Antichain, ChangeBatch, Timestamp as TimelyTimestamp};
+use timely::PartialOrder;
+use tokio::sync::{mpsc, oneshot};
+use tracing::{debug, info, trace, warn};
+
+use crate::controller::{
+    CollectionDescription, DataSource, DataSourceOther, PersistEpoch, StorageMetadata, StorageTxn,
+};
+
+/// An abstraction for keeping track of storage collections and managing access
+/// to them.
+///
+/// Responsibilities:
+///
+/// - Keeps a critical persist handle for holding the since of collections
+///   where it need to be.
+///
+/// - Drives the since forward based on the upper of a collection and a
+///   [ReadPolicy].
+///
+/// - Hands out [ReadHold] that prevent a collection's since from advancing
+///   while it needs to be read at a specific time.
+#[async_trait(?Send)]
+pub trait StorageCollections: Debug {
+    type Timestamp: TimelyTimestamp;
+
+    /// On boot, seed metadata/state.
+    async fn initialize_state(
+        &mut self,
+        txn: &mut dyn StorageTxn<Self::Timestamp>,
+        init_ids: BTreeSet<GlobalId>,
+        drop_ids: BTreeSet<GlobalId>,
+    ) -> Result<(), StorageError<Self::Timestamp>>;
+
+    /// Resets the txns system to a set of invariants necessary for correctness.
+    ///
+    /// Must be called on boot before create_collections or the various appends.
+    /// This is true _regardless_ of whether the persist-txn feature is on or
+    /// not. See the big comment in the impl of the method for details. Ideally,
+    /// this would have just been folded into `Controller::new`, but it needs
+    /// the timestamp and there are boot dependency issues.
+    ///
+    /// TODO: This can be removed once we've flipped to the new txns system for
+    /// good and there is no possibility of the old code running concurrently
+    /// with the new code.
+    async fn init_txns(
+        &mut self,
+        init_ts: Self::Timestamp,
+    ) -> Result<(), StorageError<Self::Timestamp>>;
+
+    /// Marks the end of any initialization commands.
+    ///
+    /// The implementor may wait for this method to be called before
+    /// implementing prior commands, and so it is important for a user to invoke
+    /// this method as soon as it is comfortable. This method can be invoked
+    /// immediately, at the potential expense of performance.
+    fn initialization_complete(&mut self);
```

Comment:
In my mind, any calls from the adapter/Coordinator to `StorageCollections` are commands, because you can also imagine them having a channel as interface between them. So I kept the wording from `StorageController`.

I'll now remove this method from the trait altogether, because it's existence and documentation confused both of you. 😅